# Introduction

In today's digital age, individuals are often overwhelmed by the sheer volume of available books. A book recommender system aims to alleviate this information overload by suggesting books that are likely to be of interest to a particular user. By analyzing user preferences, reading history, and book characteristics, these systems enhance the book discovery process, improve user engagement with online book platforms, and ultimately drive sales and readership. This project focuses on developing a book recommender system using three distinct datasets: books_df containing book information, ratings_df detailing user ratings, and user_df providing user demographics.

## Business Understanding

The development of an effective book recommender system holds significant value for various stakeholders:

- E-commerce Platforms and Online Bookstores: Recommender systems can drive sales by suggesting relevant books to customers, increasing the likelihood of purchases. They also enhance user experience, leading to greater customer satisfaction and loyalty.

* Libraries and Educational Institutions: These systems can help patrons discover new titles aligned with their interests or academic needs, fostering a more engaging and enriching experience.


+ Publishers and Authors: Understanding user preferences can provide valuable insights into market trends, potentially influencing publishing decisions and marketing strategies.

* Individual Users: The primary benefit for users is the ability to effortlessly discover books they are likely to enjoy, saving time and expanding their literary horizons.

By leveraging the data within the provided datasets, this project aims to build a system that accurately predicts user preferences and delivers personalized book recommendations, thereby addressing a crucial need in the book discovery ecosystem.

## Problem Statement

The core problem this project seeks to address is the challenge users face in discovering books that align with their individual tastes from a vast and ever-expanding catalog. Without an effective recommendation system, users may struggle to find new authors, genres, or titles they would enjoy, potentially leading to a less engaging and fulfilling reading experience. This project aims to develop a data-driven solution that can predict user preferences based on their past interactions and the characteristics of books, thereby providing personalized and relevant recommendations.

## Objectives

The primary objectives of this project are to:

* Develop a book recommender system: Implement one or more recommendation algorithms (e.g., collaborative filtering, content-based filtering, or hybrid approaches).

* Evaluate the performance of the recommender system: Use appropriate metrics to assess the accuracy and effectiveness of the developed model(s).

* Provide actionable insights: Based on the analysis and model results, offer insights into user preferences and potential strategies for book recommendations.

## Data Limitations

Before embarking on this project, it's important to consider potential limitations of the data:

* Data Sparsity: The ratings_df might suffer from sparsity, meaning that most users have only rated a small fraction of the available books. This can pose challenges for collaborative filtering techniques.

* Cold Start Problem: New users or new books with no or very few ratings will be difficult to recommend using collaborative filtering.

* Data Bias: The data might reflect biases present in the user base or the book catalog. For example, certain genres or authors might be over-represented.

* Data Quality: The datasets might contain inconsistencies, errors, or missing values that need to be addressed during preprocessing.

* Implicit vs. Explicit Feedback: The ratings_df likely contains explicit feedback (numerical ratings). However, other forms of implicit feedback (e.g., browsing history, purchase history) might not be available, potentially limiting the richness of user preference data.

* Evolution of Preferences: User preferences can change over time, which might not be captured effectively by static datasets.


# Data Understanding

### Loading Data

In [70]:
import pandas as pd
import numpy as np

# Load the dataset
books_df = pd.read_csv(
    '../books_df.csv',
    sep=';',                   # Semicolon separator
    quotechar='"',             # Handles quoted text properly
    encoding='latin1',         # Supports special characters
    on_bad_lines='skip',        # Skips problematic lines
    engine='python'  # Use python engine for more flexibility
)

ratings_df = pd.read_csv(
    '../ratings_df.csv',
    sep=';',                   # Semicolon separator
    quotechar='"',             # Handles quoted text properly
    encoding='latin1',         # Supports special characters
    on_bad_lines='skip',        # Skips problematic lines
    engine='python'  # Use python engine for more flexibility
)

user_df = pd.read_csv(
    '../user_df.csv',
    sep=';',                   # Semicolon separator
    quotechar='"',             # Handles quoted text properly
    encoding='latin1',         # Supports special characters
    on_bad_lines='skip',        # Skips problematic lines
    engine='python'  # Use python engine for more flexibility
)

In [71]:
books_df.head(2)

ISBN           Book-Title           Book-Author  Year-Of-Publication  \
0  0195153448  Classical Mythology    Mark P. O. Morford                 2002   
1  0002005018         Clara Callan  Richard Bruce Wright                 2001   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...

In [72]:
#summary of dataframe
books_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 270491 entries, 0 to 270490
Data columns (total 8 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   ISBN                 270491 non-null  object
 1   Book-Title           270491 non-null  object
 2   Book-Author          270489 non-null  object
 3   Year-Of-Publication  270491 non-null  int64 
 4   Publisher            270489 non-null  object
 5   Image-URL-S          270491 non-null  object
 6   Image-URL-M          270491 non-null  object
 7   Image-URL-L          270491 non-null  object
dtypes: int64(1), object(7)
memory usage: 16.5+ MB


In [73]:
# check dataframe dimension
books_df.shape

(270491, 8)

In [74]:
ratings_df.head(2)

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5

In [75]:
#summary of dataframe
ratings_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1149766 entries, 0 to 1149765
Data columns (total 3 columns):
 #   Column       Non-Null Count    Dtype 
---  ------       --------------    ----- 
 0   User-ID      1149766 non-null  int64 
 1   ISBN         1149766 non-null  object
 2   Book-Rating  1149766 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 26.3+ MB


In [76]:
# check dataframe dimension
ratings_df.shape

(1149766, 3)

In [77]:
user_df.head(2)

User-ID                   Location   Age
0        1         nyc, new york, usa   NaN
1        2  stockton, california, usa  18.0

In [78]:
#summary of dataframe
user_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278177 entries, 0 to 278176
Data columns (total 3 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   User-ID   278177 non-null  int64  
 1   Location  278177 non-null  object 
 2   Age       167669 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 6.4+ MB


In [79]:
# check dataframe dimension
user_df.shape

(278177, 3)

## Data Cleaning

In [80]:
# List of dataframes with names for clarity
dfs = [("Books", books_df), ("Ratings", ratings_df), ("Users", user_df)]

# Loop through and check missing values
for name, df in dfs:
    print(f"\nMissing values in {name} dataset:")
    missing = df.isnull().sum()
    missing = missing[missing > 0]
    if not missing.empty:
        print(missing)
    else:
        print("No missing values.")



Missing values in Books dataset:
Book-Author    2
Publisher      2
dtype: int64

Missing values in Ratings dataset:
No missing values.

Missing values in Users dataset:
Age    110508
dtype: int64


* Books dataset has missing values in Book-Author and Publisher. 
* Users dataset has missing values in Age.



## Handling missing values

We will drop the missing values in books dataset because the number is insignificant.

In [81]:
#Dropping missin values of Books data as they are negligible
books_df.dropna(inplace=True)

In [82]:
books_df.isna().sum()

ISBN                   0
Book-Title             0
Book-Author            0
Year-Of-Publication    0
Publisher              0
Image-URL-S            0
Image-URL-M            0
Image-URL-L            0
dtype: int64

In [83]:
#Fill missing Values in Age Column with Median
user_df['Age'] =user_df['Age'].fillna(user_df['Age'].median())

In [84]:
user_df.isna().sum()

User-ID     0
Location    0
Age         0
dtype: int64

In [85]:
#check summary statistics
user_df['Age'].describe()

count    278177.000000
mean         33.658624
std          11.284321
min           0.000000
25%          29.000000
50%          32.000000
75%          35.000000
max         244.000000
Name: Age, dtype: float64

Minimum age 0 and max age 244 indicates outliers exist in age

In [86]:
#Capping the outlier rows with Percentiles
upper_lim = user_df['Age'].quantile(.95)
lower_lim = user_df['Age'].quantile(.05)
user_df.loc[(user_df["Age"] > upper_lim),"Age"] = upper_lim
user_df.loc[(user_df["Age"] < lower_lim),"Age"] = lower_lim
     

In [87]:
#recheck age summary
user_df['Age'].describe()

count    278177.000000
mean         33.402449
std           9.522502
min          18.000000
25%          29.000000
50%          32.000000
75%          35.000000
max          56.000000
Name: Age, dtype: float64

### Checking for duplicates

In [88]:

# Loop through and check for duplicate rows
for name, df in dfs:
    duplicates = df.duplicated().sum()
    print(f"{name} dataset has {duplicates} duplicate rows.")


Books dataset has 0 duplicate rows.
Ratings dataset has 0 duplicate rows.
Users dataset has 0 duplicate rows.


Drop unnecessary column in books dataset

In [89]:
# Drop image URLs columns
books_df = books_df.drop(columns=['Image-URL-S', 'Image-URL-M', 'Image-URL-L'])


In [90]:
books_df.head(2)

ISBN           Book-Title           Book-Author  Year-Of-Publication  \
0  0195153448  Classical Mythology    Mark P. O. Morford                 2002   
1  0002005018         Clara Callan  Richard Bruce Wright                 2001   

                 Publisher  
0  Oxford University Press  
1    HarperFlamingo Canada